lib/kbuildbarn/exec/bb_clientd.go

```diff
@@ -0,0 +1,109 @@
+		// Make sure outputBase exists and then start bb_clientd.
+		// If there was some other error, clean out outputBase before starting bb_clientd.
+		if !errors.Is(err, os.ErrNotExist) {
+			if cleanErr := cleanOutputBase(outputBase); cleanErr != nil {
+				return nil, fmt.Errorf("failed to clean bb_clientd dir recovering from discovery error %v; clean error: %w", err, cleanErr)
+			}
+		}
+		if initErr := initOutputBase(outputBase); initErr != nil {
+			return nil, fmt.Errorf("failed to init bb_clientd output base after failing to find pidfile: %w", err)
+		}
+	}
+
+	// If exists and PID is PID of a running bb_clientd process, assume bb_clientd
+	// is running with no error
+	process, err := ps.FindProcess(pid)
+	if err != nil {
+		return nil, fmt.Errorf("unexpected error looking up bb_clientd process %d by ID: %w", pid, err)
+	}
+	if process != nil && process.Executable() == clientExecutableName {
+		return &Client{pid: pid, outputBase: outputBase}, nil
+	}
+	// Either the process wasn't found, or it wasn't bb_clientd; in either case,
+	// start a new bb_clientd instance and (over)write the pidfile.
+	pid, err = startClient(outputBase)
+	if err != nil {
+		return nil, fmt.Errorf("error starting bb_clientd: %w", err)
+	}
+	err = writePidfile(pidPath, pid)
+	if err != nil {
+		return nil, fmt.Errorf("error recording PID of bb_clientd instance to %q: %w", pidPath, err)
+	}
+	return &Client{pid: pid, outputBase: outputBase}, nil
+}
+
+func readPidfile(path string) (int, error) {
+	contents, err := os.ReadFile(path)
+	if err != nil {
+		return 0, fmt.Errorf("failed to read pidfile %q: %w", path, err)
+	}
+	i, err := strconv.ParseInt(string(contents), 10, 32)
+	if err != nil {
+		return 0, fmt.Errorf("failed to convert pid %q to a number: %w", string(contents), err)
+	}
+	return int(i), nil
+}
+
+func writePidfile(path string, pid int) error {
+	err := os.WriteFile(path, []byte(strconv.FormatInt(int64(pid), 10)), 0644)
+	if err != nil {
+		return fmt.Errorf("failed to write pid %d to file %q: %w", pid, path, err)
+	}
+	return nil
+}
+
+func startClient(outputBase string) (int, error) {
+	// Template and spill the config under outputBase
+	configPath := filepath.Join(outputbase, clientConfigSuffix)
+	// TODO: templating
+	// Start the client pointing at the config
+	cmd = exec.Command("bb_clientd", configPath)
+	err := cmd.Start()
+	if err != nil {
+		return 0, fmt.Errorf("failed to start bb_clientd: %w", err)
+	}
+	return cmd.Process.Pid, nil
+}
+
+func cleanOutputBase(outputBase string) error {
+	if err := os.RemoveAll(outputBase); err != nil {
+		return fmt.Errorf("failed to delete output base %q: %w", outputBase, err)
+	}
+	return initOutputBase(outputBase)
+}
+
+func initOutputBase(outputBase string) error {
+	if err := os.MkdirAll(filepath.Join(outputBase, clientBaseSuffix), 0755); err != nil {
+		return fmt.Errorf("failed to create output base %q: %w", outputBase, err)
+	}
+	return nil
```

Comment:
More folders need to be created here - bb_clientd gets mad if many of its internal folders as set in the config are not present on startup.